### PR TITLE
Fixed DCE of call nodes

### PR DESCRIPTION
**Pull request by rartemev**
Fixed DCE of call node which modifies stack level.
The possible situation is when there is an edge from call's block to throw helper block which depends on correct stack level. Removing call node which affects stack level we should guarantee that stack will be unwound correct.

The fix checks if there is any throwable node between eliminating call and its `PUTARG_STK` operands and set up flag whether frame pointer is required.

### CODE REVIEW

**Comment by rartemev** [comment]
@dotnet/jit-contrib PTAL

**Comment by sandreenko** [comment]
The Issue is DevDiv_524312.

File: src/jit/liveness.cpp [hunk 1]
```diff
@@ -1977,6 +1977,44 @@ void Compiler::fgComputeLife(VARSET_TP&       life,
     }
 }
 
+static bool hasAnyThrowableNodes(Compiler* compiler, BasicBlock* block, GenTree** disabledNodes, int nodeCount)
```
Comment (sandreenko) [inline src/jit/liveness.cpp]: From CLR JIT Coding Conventions:
> Functions should be named in PascalCasing format:
> The first letter of all words (including the first) should be upper case. The other letters should be lower case.

Also include a header.

File: src/jit/liveness.cpp [hunk 2]
```diff
@@ -2006,7 +2044,31 @@ void Compiler::fgComputeLifeLIR(VARSET_TP& life, BasicBlock* block, VARSET_VALAR
                     JITDUMP("Removing dead call:\n");
                     DISPNODE(call);
 
-                    node->VisitOperands([](GenTree* operand) -> GenTree::VisitResult {
+                    int       nodeCount     = 0;
+                    GenTree** disabledNodes = nullptr;
+
+                    if (call->fgArgInfo->HasStackArgs())
```
Comment (sandreenko) [inline src/jit/liveness.cpp]: Why do you need these iteration? Does it only determinate `disabledNodes` size? Why can't you use `stack` or `vector` instead?

File: src/jit/liveness.cpp [hunk 3]
```diff
@@ -2006,7 +2044,31 @@ void Compiler::fgComputeLifeLIR(VARSET_TP& life, BasicBlock* block, VARSET_VALAR
...
+                    {
+                        if ((block->bbFlags & BBF_HAS_THROW_HELPER_EDGES) != 0)
```
Comment (sandreenko) [inline src/jit/liveness.cpp]: you can also check that `codeGen->isFramePointerRequire ()== false`.

File: src/jit/liveness.cpp [hunk 4]
```diff
@@ -2016,13 +2057,24 @@ void Compiler::fgComputeLifeLIR(VARSET_TP& life, BasicBlock* block, VARSET_VALAR
                         // these nodes.
```
Comment (sandreenko) [inline src/jit/liveness.cpp]: @BruceForstall do you know should `PUTARG_SPLIT` be here as well? I suppose it changes stack level and we should check it too.

File: src/jit/liveness.cpp [hunk 5]
```diff
@@ -1977,6 +1977,45 @@ void Compiler::fgComputeLife(VARSET_TP&       life,
...
+
+    LIR::Range& blockRange  = LIR::AsRange(block);
+    GenTree*    currentNode = blockRange.LastNode();
+    GenTree*    endNode     = blockRange.FirstNonPhiNode()->gtPrev;
+
+    size_t left = disabledNodes.size();
+
+    while (currentNode != endNode)
+    {
+        for (size_t i = 0; i < disabledNodes.size(); ++i)
```
Comment (sandreenko) [inline src/jit/liveness.cpp]: I understand that `disabledNodes.size()` is small, but I would prefer `set` or `hash map` here, but it is up to you.

**Comment by sandreenko** [comment]
I expect many asm diffs, because we change frame from `rsp` to `rbp` each time when we delete a call, that has `put_stk_arg` in another block. I do not understand why we hit the issue only on one test.

```
block1
  1 put_arg_stk
block2
  2 call <- delete this call, can't prove that there is no throw nodes between 1 and 2, so set rbp required.
```

Would be nice to look at spmi diffs to understand it or maybe somebody has a clue about it.

File: src/jit/liveness.cpp (hunk 5 above)
Comment (rartemev) [reply src/jit/liveness.cpp]: both set and hashmap consume much more memory. I think in case of small 'n' it's better to go with flat data structures.

**Comment by rartemev** [comment]
@sandreenko 
I am not sure that cases you described are even possible but to prevent any failures there I added `return true` in the end of `HasAnyThrowableNodes` function to handle such situation. This statement means there is a `PUT_ARG` node somewhere outside the current block so we'd rather use `rbp`.

**Comment by rartemev** [comment]
@sandreenko I made spmidiff and there is no difference.

**Pull request merged**